### ISSUE TITLE
4: Neural Net Foundations (Practical Deep Learning for Coders)

### ISSUE THREAD

**Issue by JeremyLoh** [JeremyLoh/ml#8]
https://course.fast.ai/Lessons/lesson3.html

Today we’ll be learning about the mathematical foundations of deep learning: Stochastic gradient descent (SGD), and the flexibility of linear functions layered with non-linear activation functions. We’ll be focussing particularly on a popular combination called the Rectified linear function (ReLU).

**Chapter 4 is the point where the course loses the most people**

https://machine-learning.paperspace.com/

# Kaggle notebook for practice
### https://www.kaggle.com/code/jeremyloh42/chapter-4-deep-learning-for-coders

# Tasks
- [x] 1:22:35 / 1:30:24 Lesson 3: Practical Deep Learning for Coders 2022 https://www.youtube.com/watch?v=hBBOjCiFcuo
- [ ] Mastering Derivatives for machine learning https://towardsai.net/p/machine-learning/mastering-derivatives-for-machine-learning
- [ ] Transpose of matrix https://www.cuemath.com/algebra/transpose-of-a-matrix/
- [ ] Read Chapter 4 of book (Deep Learning for Coders with Fastai and PyTorch) - Start Page ??? / 1028
    - Current Page: ???

![Image](https://user-images.githubusercontent.com/38811217/268298860-8fab6223-20d0-4162-bdff-cfa949c63f5b.png)
![Image](https://user-images.githubusercontent.com/38811217/268298740-3e154881-cfda-4072-8677-777dde5b34c5.png)

# https://stackoverflow.com/questions/48001598/why-do-we-need-to-call-zero-grad-in-pytorch

1:24:44
Getting started with nlp for absolute beginners - https://www.kaggle.com/code/jhoward/getting-started-with-nlp-for-absolute-beginners
- NLP (Natural language processing) is about making predictions using text data. The text data mostly is in form of prose (e.g. plain english text)
- Most widely useful application of NLP is calssification, take a document (e.g. book, wikipedia page) and automatically figure out a category for it. 
    - sentimental analysis (e.g. are people saying good or bad things about your product) 
    - author identification (who wrote this document)
- We are going to use a different library, Hugging Face transformers, not as high level as fastai. Has good models in it, good techniques in the library

13:49
- Which image models are best? - https://www.kaggle.com/code/jhoward/which-image-models-are-best

We only care about 3 things:
1. How fast is the model
2. How accurate is the model
3. How much memory does the model use

23:52
How does a neural net really work?
https://www.kaggle.com/code/jhoward/how-does-a-neural-net-really-work
Machine learning models are things that fit functions to data
We start with an infinitely flexible function, neural network, and we get it to do a particular thing

31:25
A way to measure how good our model is -> done through a **loss function**
Common loss function is mean squared error (mse), (predictions - actuals)**2 and get the mean

33:50
Automating finding the best possible value for loss
For each parameter, when we move it, does the loss improve?
Much faster way is to calculate the derivative. derivative measures the rate of change of a function at a particular point. This information can be used to optimize functions, find local minima and maxima, and more.
https://towardsai.net/p/machine-learning/mastering-derivatives-for-machine-learning

Pytorch can calculate derivative for us.
Pytorch .mean() returns a 1D tensor. A tensor is something that works with numbers, list of numbers, tables of numbers etc. Everything in Pytorch is a tensor
1D Tensor is a rank-1 tensor. e.g.
`abc = torch.tensor([1.5, 1.5, 1.5])`
`abc.requires_grad_()` is used to tell Pytorch to calculate the gradient for these numbers whenever we use them in a calculation. It is a flag on the tensor

50:20
Try to use faster models first for experimentation (e.g. resnet34), do data cleaning, data augmentation, before changing to a faster model. Sometimes the result is good enough, for the use case we are doing

1:04:40
Dependent variable is the thing that we are trying to predict, e.g. titanic, did the person survive?
https://www.kaggle.com/competitions/titanic/data?select=train.csv
Data dictionary helps to explain data columns and their meaning, for the given dataset
- Pclass - class of boat
- SibSp - siblings in family

Remove rows that have blank value for "age" in the titanic train dataset, and "embarked"

1:07:35
After data cleanup, we need to multiply each column by some coefficient.
- How do we multiply the word "male" by a coefficient? -> We need to convert them into numbers

![Image](https://user-images.githubusercontent.com/38811217/267993270-332a00ef-b91c-46a0-bfba-26cdb878e791.png)

1:08:40
Normalizing data where columns have huge numbers, e.g. age has 2 to 80, we can do age / MAX(age), 

1:11:00
fare => LOG10(fare + 1)
Money has lots of small numbers and some big ones
If you take the LOG of something with a huge distribution, it becomes something that is more evenly distributed

1:13:00
y = mx+b for linear equation but in the excel, we don't have a plus at the end?
we have a col of data that is all 1, that parameter is the constant term. No need to have a constant term (b), everyone does it

![Image](https://user-images.githubusercontent.com/38811217/268014219-f0a4fd63-374d-4389-b387-a7ecde482c2e.png)

1:17:00
Doing ReLU (Rectified Linear Unit), if value < 0 ? 0 : value

![Image](https://user-images.githubusercontent.com/38811217/268027796-93b4a168-74df-4fb6-b759-243d757b8a6c.png)

Doing predictions from adding the 2 ReLU together
![Image](https://user-images.githubusercontent.com/38811217/268028210-03eefc37-50a6-49e2-a758-d094cec60937.png)

1:18:35
Using matrix multiplication, parameters are transposed

![Image](https://user-images.githubusercontent.com/38811217/268029979-8f7b1e36-2981-49e6-acbc-281514d016f1.png)

1:22:45
Why is there no Pclass_3 categorical variable? That column has 3 possible values, so we only need 2 variables to infer all combinations.
For categorical variable with N levels, it needs N - 1 variables
